tensorboard/webapp/widgets/line_chart_v2/lib/renderer/threejs_renderer.ts

```diff
@@ -32,13 +37,97 @@ function createOpacityAdjustedColor(hex: string, opacity: number): THREE.Color {
   return new THREE.Color(newD3Color.brighter(1.6 - opacity).hex());
 }
 
+enum CacheType {
+  CIRCLE,
+  LINE,
+  TRIANGLE,
+}
+
 interface LineCacheValue {
-  type: 'line';
-  obj3d: THREE.Object3D;
+  type: CacheType.LINE;
+  obj3d: THREE.Line;
   data: Polyline;
 }
 
-type CacheValue = LineCacheValue;
+interface TriangleCacheValue {
+  type: CacheType.TRIANGLE;
+  obj3d: THREE.Mesh;
+  data: Point;
+}
+
+interface CircleCacheValue {
+  type: CacheType.CIRCLE;
+  obj3d: THREE.Mesh;
+  data: {loc: Point; radius: number};
+}
+
+type CacheValue = LineCacheValue | TriangleCacheValue | CircleCacheValue;
+
+/**
+ * Updates BufferGeometry with Float32Array that denotes flattened array of Vec2 (<x, y>).
+ */
+function updateGeometryWithVec2Array(
+  geometry: THREE.BufferGeometry,
+  flatVec2: Float32Array
+) {
+  const length = flatVec2.length / 2;
+  let positionAttributes = geometry.attributes[
+    'position'
+  ] as THREE.BufferAttribute;
+  if (!positionAttributes || positionAttributes.count !== length * 3) {
+    positionAttributes = new THREE.BufferAttribute(
+      new Float32Array(length * 3),
+      3
+    );
+    geometry.addAttribute('position', positionAttributes);
+  }
+  const values = positionAttributes.array as number[];
+  for (let index = 0; index < length; index++) {
+    values[index * 3] = flatVec2[index * 2];
+    values[index * 3 + 1] = flatVec2[index * 2 + 1];
+  }
+  positionAttributes.needsUpdate = true;
+  geometry.setDrawRange(0, length * 3);
+  // Need to update the bounding sphere so renderer does not skip rendering
+  // this object because it is outside of the camera viewpoint (frustum).
+  geometry.computeBoundingSphere();
+}
+
+/**
+ * Updates an THREE.Object3D like object with geometry and material. Returns true if
+ * geometry is updated with the new data but returns false if we can minimally change the
+ * material to make an update.
+ */
+function updateObject(
+  object: THREE.Mesh | THREE.Line,
+  updateGeometry: (geometry: THREE.BufferGeometry) => THREE.BufferGeometry,
+  materialOption: {visible: boolean; color: string; opacity?: number}
+): boolean {
+  const {visible, color, opacity} = materialOption;
+
+  if (Array.isArray(object.material)) {
+    throw new Error('Invariant error: only expect one material on an object');
+  }
+
+  const material = object.material as THREE.MeshBasicMaterial;
+  if (material.visible !== visible) {
+    material.visible = visible;
+    material.needsUpdate = true;
+  }
+
+  if (!visible) return false;
+
+  const newColor = createOpacityAdjustedColor(color, opacity ?? 1);
+
+  updateGeometry(object.geometry as THREE.BufferGeometry);
+  const currentColor = material.color;
+  if (!currentColor.equals(newColor)) {
+    material.color.set(newColor);
+    material.needsUpdate = true;
+  }
+
+  return true;
```

Comment:
Seems like there are some cases where updateGeometry might not make any changes to object.geometry but we still return true.  Is that intentional/desirable. It seems to contradict to the documentation?

Reply:
It returns true when `updateGeometry` is invoked. I can make the users remember the invocation state in the callback but I thought it would read messier that way. Made some update to the document to mitigate the confusion. WDYT?